File: src/application/export.ts
// Was zwei Exporte teilen: wohin geschrieben wird und wie die Datei heisst.
//
// **Warum das hier steht und nicht im Konfigurationsexport.** Bis 2026-08-30 gab es genau
// einen Export, und beides lag bei ihm. Seit der Bestandsexport daneben steht, benutzen
// zwei Use-Cases denselben Port — und ein Port, den zwei benutzen, gehoert keinem von
// beiden. Ohne diesen Schritt haette `bestandsexport` von `konfiguration` importieren
// muessen, obwohl die beiden fachlich nichts miteinander zu tun haben.
//
// **Der Unterschied zwischen den beiden Dateien ist keine Formsache**, und deshalb faellt
// er hier zusammen mit dem Namen: `konfiguration-…` sagt, wie der Haushalt ORDNET, und
// darf weitergegeben werden; `bestand-…` sagt, was in ihm passiert ist, und darf es nicht.
// Wer eine dritte Art anlegt, entscheidet mit dem Praefix, in welche der beiden Klassen
// sie faellt — und die Antwort gehoert in denselben Kommentar wie diese hier.

/**
 * Wohin eine Exportdatei geht.
 *
 * Der Port kennt keinen Pfad, nur einen Dateinamen: WO exportiert wird, entscheidet der
 * Adapter (und dahinter das Rust-Kommando), nicht der Use-Case. Ein Use-Case, der ein
 * Verzeichnis benennt, haette eine Meinung ueber das Dateisystem — und die gehoert nicht
 * in die Anwendungsschicht.
 */
export interface ExportZiel {
  /** Schreibt die Datei und meldet, wo sie gelandet ist. */
  schreiben(name: string, inhalt: string): Promise<string>;
}

/**
 * Was fuer ein Export es ist — und damit, wie die Datei heisst.
 *
 * Zwei Werte, zwei Zusicherungen: `konfiguration` traegt keine Zahlung und darf deshalb
 * das Haus verlassen, `bestand` traegt IBANs, Salden und jeden Verwendungszweck. Der Name
 * ist die einzige Stelle, an der man einer Datei von aussen ansieht, welche der beiden man
 * vor sich hat — im Dateimanager, im Anhang einer Mail, in einem Ordner voller Exporte.
 */
export type Exportart = "konfiguration" | "bestand";

/**
 * Der Dateiname zu einer Art, einem Tag und einem Bestand.
 *
 * **Die Bestandskennung ist nicht Zierrat.** Der echte Bestand und der Spielstand liegen
 * in zwei Dateien, aber im SELBEN App-Datenverzeichnis — der Identifier trennt sie nicht
 * (siehe `datenbankdatei.ts`). Ohne Kennung schreiben beide Apps denselben Namen, und ein
 * Export aus der installierten App ueberschriebe den des Spielstands wortlos. Die beiden
 * sehen von aussen gleich aus; welcher gemeint war, wuesste danach niemand mehr.
 *
 * Ein Export je Art, Tag und Bestand, der neuere ersetzt den aelteren. Bei einer
 * Momentaufnahme ist das richtig — anders als bei den Sicherungen, wo der ALTE Stand der
 * wertvolle ist.
 */
export function exportDateiname(art: Exportart, erzeugt: Date, bestand: string): string {
  const kennung = bestand.replace(/\.db$/, "");
  return `${art}-${kennung}-${erzeugt.toISOString().slice(0, 10)}.json`;
}

File: src-tauri/src/export.rs
// Eine Datei aus der App heraus — der einzige Schreibweg nach draussen.
//
// **Warum ein eigenes Kommando und kein `<a download>`.** Der naheliegende Weg im Web ist
// ein Blob mit `download`-Attribut. Im WKWebView, den Tauri auf macOS benutzt, ist das
// unzuverlaessig: der Download landet je nach Fassung nirgends oder wortlos im
// Papierkorb-Verzeichnis des Webviews. Ein Export, von dem man nicht weiss, wo er liegt,
// ist keiner. Dieselbe Ueberlegung wie beim Datenbankzugang, der aus demselben Grund ueber
// eigene Kommandos laeuft statt ueber ein Plugin.
//
// **Wohin geschrieben wird, entscheidet NICHT der Aufrufer.** Ziel ist der
// DOWNLOAD-Ordner des Nutzers, und der Name muss ein einfacher Dateiname sein. Ein
// Webview, der irgendwohin schreiben darf, ist ein Webview, der ueberall hinschreiben
// kann — und was den Bestand liest, ist derselbe Prozess. Die Abwehr bleibt damit
// dieselbe; nur das Verzeichnis ist ein anderes.
//
// **Bis 2026-09-22 war es `<App-Datenverzeichnis>/export/`**, und das war die falsche
// Haelfte der richtigen Ueberlegung. Der Ort war sicher und unauffindbar: eine Exportdatei
// ist dazu da, WEITERGEGEBEN zu werden — an ein Tabellenprogramm, an den naechsten Rechner,
// in den Anhang einer Mail —, und dafuer muss man sie greifen koennen. Wer sie im
// App-Datenverzeichnis sucht, sucht in einem versteckten Pfad, dessen Namen nur die Karte
// kennt. Der Download-Ordner ist der Ort, an dem jedes andere Programm seine Dateien
// ablegt, und er gehoert dem Nutzer, nicht uns.
//
// Faellt er aus (kein Download-Ordner auffindbar), bleibt der alte Weg als Rueckfall — ein
// Export, der stattfindet und an einem unbequemen Ort landet, ist besser als keiner.
//
// **Es ueberschreibt.** Ein zweiter Export desselben Tages ersetzt den ersten. Das ist bei
// einer Momentaufnahme richtig: der neuere Stand ist der bessere, und eine Datei je Klick
// waere ein Verzeichnis, das niemand aufraeumt. Der Unterschied zu den Sicherungen ist die
// Absicht — dort ist der ALTE Stand der wertvolle.

use std::path::PathBuf;

use tauri::{AppHandle, Manager};

/// Das Verzeichnis, in das exportiert wird — angelegt, falls es fehlt.
///
/// Der Download-Ordner ist der Ort, an dem der Nutzer eine Datei sucht; er existiert auf
/// jedem System, das uns interessiert, und muss deshalb im Normalfall nicht angelegt
/// werden. Der Rueckfall ins App-Datenverzeichnis ist fuer den Rest.
fn exportverzeichnis(app: &AppHandle) -> Result<PathBuf, String> {
    if let Ok(download) = app.path().download_dir() {
        if std::fs::create_dir_all(&download).is_ok() {
            return Ok(download);
        }
    }
    let basis = app
        .path()
        .app_data_dir()
        .map_err(|e| format!("Kein App-Datenverzeichnis: {e}"))?
        .join("export");
    std::fs::create_dir_all(&basis).map_err(|e| format!("Exportverzeichnis fehlt: {e}"))?;
    Ok(basis)
}

/// Prueft den Namen und weist alles ab, was wie ein Pfad aussieht.
///
/// Dieselbe Abwehr wie in `datenbank::datei_im_datenverzeichnis`, und aus demselben Grund
/// wortgleich streng: ein zurechtgebogener Pfad ist die Sorte Abwehr, die beim naechsten
/// Sonderzeichen nicht mehr traegt.
fn geprueft(name: &str) -> Result<&str, String> {
    if name.is_empty()
        || name.contains('/')
        || name.contains('\\')
        || name.contains("..")
        || name.starts_with('.')
    {
        return Err(format!("'{name}' ist kein einfacher Dateiname."));
    }
    Ok(name)
}

/// Schreibt `inhalt` in den Download-Ordner und meldet den vollen Pfad.
///
/// Der Pfad geht zurueck an die Oberflaeche, obwohl er jetzt an einem bekannten Ort liegt:
/// „im Download-Ordner" ist eine Auskunft, der genaue Name ist die, mit der man die Datei
/// auch findet, wenn dort dreihundert andere liegen.
#[tauri::command]
pub async fn export_schreiben(app: AppHandle, name: String, inhalt: String) -> Result<String, String> {
    let datei = exportverzeichnis(&app)?.join(geprueft(&name)?);
    std::fs::write(&datei, inhalt).map_err(|e| format!("Export nicht schreibbar: {e}"))?;
    Ok(datei.to_string_lossy().into_owned())
}

#[cfg(test)]
mod tests {
    use super::geprueft;

    #[test]
    fn nur_einfache_dateinamen_kommen_durch() {
        for gut in ["kategorien.json", "export-2026-08-30.json", "a"] {
            assert!(geprueft(gut).is_ok(), "'{gut}' sollte durchgehen.");
        }
        // Der Webview darf sich das Ziel nicht selbst aussuchen: er liest denselben
        // Bestand, den er sonst nirgendwohin tragen kann (siehe CSP).
        for boese in ["../heimlich.json", "/etc/passwd", "unter/pfad.json", ".versteckt", ""] {
            assert!(geprueft(boese).is_err(), "'{boese}' haette abgewiesen werden muessen.");
        }
    }
}
